standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/impl/FindColumnsWithStatsHandler.java

```diff
@@ -37,11 +37,16 @@ public class FindColumnsWithStatsHandler implements QueryHandler<List<String>> {
 
   //language=SQL
   private static final String TABLE_SELECT = "SELECT \"COLUMN_NAME\" FROM \"TAB_COL_STATS\" " +
-      "WHERE \"DB_NAME\" = :dbName AND \"TABLE_NAME\" = :tableName";
+      "INNER JOIN \"TBLS\" ON \"TAB_COL_STATS\".\"TBL_ID\" = \"TBLS\".\"TBL_ID\" " +
```

Comment:
any clue how that would affect the read performance? denormalization is used to avoid extra joins.

Reply:
The join could hit index, for example in `PARTITIONS` table it could hit the `UNIQUEPARTITION (PART_NAME, TBL_ID) ` index so it would not affect the performance.

Reply:
Chime in with some thoughts :)
Not sure if the index can be used to avoid performance degression when multi-join.
There are many read(select) operations related to statistics in Hive, especially in CBO stage. Sometimes the performance of mutli-join operation in MySQL is bad than single table operation, and the multi-join also cause mysql performance stress. 
https://github.com/apache/hive/pull/4744/files#diff-bcca13f6cc251df321e8fe80568ef0334a1d44f7e5e7ff2fcaa06ab4f05bbdf9 `MetaStoreDirectSql.java` also changed the stats related operation from single table to multi-join operation.

Can we do some performance stress tests to verify that performance won't decline. (maybe not easy to test)
Thanks.

Reply:
Found sveral similar tickets about TAB_COL_STATS and PART_COL_STATS:
https://issues.apache.org/jira/browse/HIVE-6190
https://issues.apache.org/jira/browse/HIVE-8524

Reply:
`UNIQUEPARTITION ` index is high performance, we have a table with 4 millions partitions, and the API`getPartitionByName` hitting the `UNIQUEPARTITION` index only takes around 200 milliseconds.

Also let me add a benchmark test for this.

Reply:
@zhangbutao @deniskuzZ The benchmark test result shows there was no obvious performance decline:
```bash
java -jar ./hmsbench-jar-with-dependencies.jar -H localhost --savedata /tmp/benchdata --sanitize -N 100 -N 1000 -o bench_results_direct.csv -C -d testbench_http --params=100  -E 'drop.*' -E 'renameTable.*' -E 'getTableObjectsByName.*' -E 'listTables.*' -E 'listPartitions.*' -E 'getPartitionsByNames.*' -E 'getPartitionNames.*' -E 'listPartition' -E 'getPartition' -E 'getPartitions' -E 'getPartitions.10' -E 'getPartitions.100' -E 'getPartitions.1000' -E 'addPartition' -E 'addPartitions.*' -E 'getNid' -E 'listDatabases' -E 'getTable' -E 'createTable' -E 'openTxn.*'
```
- Before this patch
```bash
Operation                      Mean     Med      Min      Max      Err%
getPartitionsStat              4.053    3.982    3.757    5.958    6.661
getPartitionsStat.100          4.522    4.429    4.120    6.520    7.953
getPartitionsStat.1000         6.976    6.808    6.493    11.46    10.39
updatePartitionsStat           7.054    7.019    6.679    8.144    3.218
updatePartitionsStat.100       1507     1478     1447     3230     12.33
updatePartitionsStat.1000      1.491e+04 1.479e+04 1.459e+04 1.661e+04 2.386
```
- After this patch
```bash
Operation                      Mean     Med      Min      Max      Err%
getPartitionsStat              4.018    3.970    3.666    5.845    6.487
getPartitionsStat.100          4.091    4.049    3.838    4.789    4.371
getPartitionsStat.1000         6.248    6.142    5.869    10.95    8.817
updatePartitionsStat           7.031    6.813    6.312    15.33    15.20
updatePartitionsStat.100       1470     1454     1427     1871     5.304
updatePartitionsStat.1000      1.468e+04 1.454e+04 1.440e+04 1.667e+04 2.231
```

Reply:
👍

Reply:
This result was in mysql, let me add the test on postgres.

Reply:
thanks @wecharyu , looks Ok

Reply:
Those tests were performed for direct SQL. What about ORM? Would it be the same?

Reply:
The ORM was similar, here is a test result in mysql ORM:
- Before this patch:
```bash
Operation                      Mean     Med      Min      Max      Err%
getPartitionsStat              7.583    7.432    5.914    11.53    13.82
getPartitionsStat.100          5.956    5.778    5.280    12.48    14.01
getPartitionsStat.1000         12.12    11.93    11.00    13.97    5.809
updatePartitionsStat           11.75    11.69    10.32    14.97    6.476
updatePartitionsStat.100       1581     1572     1510     1961     4.320
updatePartitionsStat.1000      1.579e+04 1.564e+04 1.536e+04 1.778e+04 2.859
```
- After this patch:
```bash
Operation                      Mean     Med      Min      Max      Err%
getPartitionsStat              7.323    7.341    5.533    10.45    13.74
getPartitionsStat.100          5.168    5.079    4.643    9.383    10.01
getPartitionsStat.1000         11.35    11.14    10.51    13.49    5.317
updatePartitionsStat           10.86    10.67    8.824    19.97    14.70
updatePartitionsStat.100       1537     1523     1471     1917     4.554
updatePartitionsStat.1000      1.563e+04 1.549e+04 1.517e+04 1.774e+04 2.890
```

Reply:
👍